### PR TITLE
[#181] Common base class for request response client implementations.

### PR DESCRIPTION
**Pull request by sysexcontrol**
Second refactoring for common base classes for request response APIs - focussed on the client implementation.

Signed-off-by: Karsten Frank <Karsten.Frank@bosch-si.com>

### CODE REVIEW

File: client/src/main/java/org/eclipse/hono/client/RequestResponseClient.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ * Contributors:
+ *    Bosch Software Innovations GmbH - initial creation
+ */
+package org.eclipse.hono.client;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Handler;
+
+public interface RequestResponseClient {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/RequestResponseClient.java]: what about a comment explaining the purpose of this interface?

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A Vertx-Proton based parent class for the implementation of API clients that follow the request response pattern.
+ * The class is a generic that expects two classes:
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: this explanation is good to have, but it should be done via `@param` tags.

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    /**
+     * Build a Proton message with a provided subject (serving as the operation that shall be invoked).
+     * The message can be extended by arbitrary application properties passed in.
+     *
+     * @param subject The subject system property of the message.
+     * @param appProperties The map containing arbitrary application properties.
+     *                      Maybe null if no application properties are needed.
+     * @return The Proton message constructed from the provided parameters.
+     */
+    protected Message createMessage(final String subject, final Map<String, Object> appProperties) {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: is this supposed to be overridden?

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+    protected final void sendMessage(final Message request, final Handler<AsyncResult<R>> resultHandler) {
+
+        context.runOnContext(req -> {
+            replyMap.put((String) request.getMessageId(), resultHandler);
+            sender.send(request);
+        });
+    }
+
+    public final boolean isOpen() {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: `@Override`?

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        return sender != null && sender.isOpen() && receiver != null && receiver.isOpen();
+    }
+
+    public final void close(final Handler<AsyncResult<Void>> closeHandler) {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: `@Override`?

File: client/src/main/java/org/eclipse/hono/client/impl/CredentialsClientImpl.java [hunk 6]
```diff
@@ -38,73 +39,33 @@
...
-    private static final String                  CREDENTIALS_REPLY_TO_ADDRESS_TEMPLATE = "credentials/%s/%s";
+    private static final String                  CREDENTIALS_NAME = "credentials";
 
-    private final Map<String, Handler<AsyncResult<CredentialsResult>>> replyMap = new ConcurrentHashMap<>();
-    private final String                         credentialsReplyToAddress;
+    private static final Logger                  LOG = LoggerFactory.getLogger(CredentialsClientImpl.class);
 
     private CredentialsClientImpl(final Context context, final ProtonConnection con, final String tenantId,
                                   final Handler<AsyncResult<CredentialsClient>> creationHandler) {
+        super(context,con,tenantId,creationHandler);
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/CredentialsClientImpl.java]: insert spaces after commas

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    protected final AtomicLong            messageCounter  = new AtomicLong();
+    private final String                  requestResponseAddressTemplate;
+    private final String                  requestResponseReplyToAddressTemplate;;
+
+    protected final Map<String, Handler<AsyncResult<R>>> replyMap = new ConcurrentHashMap<>();
+    protected final String                replyToAddress;
+
+    /**
+     * Get the name of the endpoint that this client targets at.
+     * @return The name of the endpoint for this client.
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: insert empty line

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+     */
+    protected abstract String getName();
+
+    /**
+     * Build a unique messageId for a request that serves as an identifier for a new message.
+     * @return The unique messageId;
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: insert empty line

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+     */
+    protected abstract String createMessageId();
+
+    /**
+     * Construct a result object that contains the answer for a message.
+     * @param status The status of the response.
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: insert empty line

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+     * @param payload The json payload of the response.
+     * @return The result object.
+     */
+    protected abstract R from(final int status, final JsonObject payload);
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: rename to `getResult`?

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+    protected abstract String getName();
+
+    /**
+     * Build a unique messageId for a request that serves as an identifier for a new message.
+     * @return The unique messageId;
+     */
+    protected abstract String createMessageId();
+
+    /**
+     * Construct a result object that contains the answer for a message.
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: *Creates a result for the status and payload from a response received from the endpoint.*?

File: core/src/main/java/org/eclipse/hono/util/RegistrationResult.java [hunk 12]
```diff
@@ -18,14 +18,9 @@
  * A container for the result returned by Hono's registration API.
  *
  */
-public final class RegistrationResult {
-
-    private final int status;
-    private final JsonObject payload;
-
+public final class RegistrationResult extends RequestResponseResult {
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/RegistrationResult.java]: do we need this concrete class?

File: core/src/main/java/org/eclipse/hono/util/CredentialsResult.java [hunk 13]
```diff
@@ -18,14 +18,9 @@
  * A container for the result returned by Hono's credentials API.
  *
  */
-public final class CredentialsResult {
-
-    private final int status;
-    private final JsonObject payload;
-
+public final class CredentialsResult extends RequestResponseResult {
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/CredentialsResult.java]: do we need this concrete class?

File: core/src/main/java/org/eclipse/hono/util/RegistrationResult.java (hunk 12 above)
Comment (sysexcontrol) [inline core/src/main/java/org/eclipse/hono/util/RegistrationResult.java]: Formally not (could be substituted by RequestResponseResult), but I thought it is better to keep a specific result class per API available to have an extension point available for each API.

I propose to keep it like that, but there are no pure technical reasons for it.
WDYT?

File: core/src/main/java/org/eclipse/hono/util/CredentialsResult.java (hunk 13 above)
Comment (sysexcontrol) [inline core/src/main/java/org/eclipse/hono/util/CredentialsResult.java]: see RegistrationResult.

**Comment by sysexcontrol** [comment]
Updated, please have a look again, thanks!

File: core/src/main/java/org/eclipse/hono/util/RegistrationResult.java (hunk 12 above)
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/RegistrationResult.java]: ok, makes sense

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java (hunk 3 above)
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: the passed in `appProperties` should be checked/filtered for compatible value types as we do in `AbstractSender.setApplicationProperties`. Maybe we can pull up that method into `AbstractHonoClient` and use it here as well?

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+ * @param <C> denotes the concrete interface for the API
+ * @param <R> denotes the concrete result container class of the API
+ *
+ * <p>
+ * Both type parameter classes have their own parent and need to be subclassed.
+ *
+ * A subclass of this class only needs to implement some abstract helper methods (see the method descriptions) and their own
+ * API specific methods. This allows for implementation classes that focus on the API specific code.
+ */
+public abstract class AbstractRequestResponseClient<C extends RequestResponseClient,R extends RequestResponseResult>
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: there is a space missing after the comma

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+            request.setBody(new AmqpValue(payload.encode()));
+        }
+        sendMessage(request, resultHandler);
+    }
+
+    protected final void sendMessage(final Message request, final Handler<AsyncResult<R>> resultHandler) {
+
+        context.runOnContext(req -> {
+            replyMap.put((String) request.getMessageId(), resultHandler);
+            sender.send(request);
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: we currently assume that we always have enough credit (which seems fair at the moment) but I wonder if (in a future PR) we should add functionality for waiting on more credits before sending out additional requests when we run out of credit ...

File: client/src/main/java/org/eclipse/hono/client/impl/CredentialsClientImpl.java [hunk 16]
```diff
@@ -38,73 +39,33 @@
  * A Vertx-Proton based client for Hono's Credentials API.
  *
  */
-public final class CredentialsClientImpl extends AbstractHonoClient implements CredentialsClient {
+public final class CredentialsClientImpl extends AbstractRequestResponseClient<CredentialsClient,CredentialsResult> implements CredentialsClient {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/CredentialsClientImpl.java]: space after comma

File: client/src/main/java/org/eclipse/hono/client/impl/RegistrationClientImpl.java [hunk 17]
```diff
@@ -17,102 +17,55 @@
...
-import io.vertx.proton.ProtonQoS;
-import io.vertx.proton.ProtonReceiver;
-import io.vertx.proton.ProtonSender;
 
 /**
  * A Vertx-Proton based client for Hono's Registration API.
  *
  */
-public final class RegistrationClientImpl extends AbstractHonoClient implements RegistrationClient {
+public final class RegistrationClientImpl extends AbstractRequestResponseClient<RegistrationClient,RegistrationResult> implements RegistrationClient {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/RegistrationClientImpl.java]: space after comma

File: core/src/main/java/org/eclipse/hono/util/RequestResponseResult.java [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+
+    protected RequestResponseResult(final int status, final JsonObject payload) {
+        this.status = status;
+        this.payload = payload;
+    }
+
+    /**
+     * @return the status
+     */
+    public int getStatus() {
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/RequestResponseResult.java]: final

File: core/src/main/java/org/eclipse/hono/util/RequestResponseResult.java [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+        return status;
+    }
+
+    /**
+     * @return the payload
+     */
+    public JsonObject getPayload() {
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/RequestResponseResult.java]: final

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java (hunk 15 above)
Comment (sysexcontrol) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: Both good points +1
- appProperties are now checked as proposed
- waiting for credits for now added as TODO (detailed javadoc added as well)

**Pull request merged**